File: packages/cc-headless/src/report-store.ts
import { S3Client, PutObjectCommand } from '@aws-sdk/client-s3';
import { getSignedUrl } from '@aws-sdk/s3-request-presigner';
import { GetObjectCommand } from '@aws-sdk/client-s3';
import { SNSClient, PublishCommand } from '@aws-sdk/client-sns';

const S3_REPORT_BUCKET = process.env.S3_REPORT_BUCKET ?? '';
const SNS_TOPIC_ARN = process.env.SNS_NOTIFICATION_TOPIC_ARN ?? '';
const PRESIGNED_URL_EXPIRY = 86400; // 24 hours

const s3 = new S3Client({});
const sns = new SNSClient({});

export async function saveReport(
  rcaId: string,
  reportMarkdown: string,
): Promise<string> {
  const key = `reports/${rcaId}.md`;
  if (!S3_REPORT_BUCKET) return key;

  await s3.send(
    new PutObjectCommand({
      Bucket: S3_REPORT_BUCKET,
      Key: key,
      Body: reportMarkdown,
      ContentType: 'text/markdown',
    }),
  );

  return key;
}

export async function sendNotification(
  rcaId: string,
  alarmName: string,
  rootCause: string,
  reportS3Key: string,
  elapsedSeconds: number,
): Promise<void> {
  if (!SNS_TOPIC_ARN) return;

  let reportUrl = `s3://${S3_REPORT_BUCKET}/${reportS3Key}`;
  if (S3_REPORT_BUCKET) {
    try {
      reportUrl = await getSignedUrl(
        s3,
        new GetObjectCommand({
          Bucket: S3_REPORT_BUCKET,
          Key: reportS3Key,
        }),
        { expiresIn: PRESIGNED_URL_EXPIRY },
      );
    } catch {
      // fall back to S3 URI
    }
  }

  const message = JSON.stringify({
    rca_id: rcaId,
    alarm_name: alarmName,
    root_cause: rootCause,
    report_url: reportUrl,
    engine: 'cc-headless',
    elapsed_seconds: elapsedSeconds,
  });

  await sns.send(
    new PublishCommand({
      TopicArn: SNS_TOPIC_ARN,
      Subject: `[RCA] ${alarmName} — Analysis Complete (cc-headless)`,
      Message: message,
    }),
  );
}

File: packages/cc-headless/src/cc_headless/report_store.py
import boto3 as _boto3
from botocore.config import Config as _Config

from cc_headless.adapters.secondary.report.s3_report_store import S3ReportStore  # noqa: F401
from cc_headless.config.settings import (  # noqa: F401
    ENGINE,
    PRESIGNED_URL_EXPIRY,
    S3_REPORT_BUCKET,
    SNS_NOTIFICATION_TOPIC_ARN,
)

_s3 = _boto3.client("s3", config=_Config(signature_version="s3v4"))
_sns = _boto3.client("sns")
_default_store = S3ReportStore(_s3, _sns)


def save_report(rca_id, report_markdown):
    return _default_store.save_report(rca_id, report_markdown)


def send_notification(
    rca_id,
    alarm_name,
    root_cause,
    report_s3_key,
    elapsed_seconds,
    *,
    playbook=None,
):
    return _default_store.send_notification(
        rca_id,
        alarm_name,
        root_cause,
        report_s3_key,
        elapsed_seconds,
        playbook=playbook,
    )
